nixos/chromium: simple option to install chromium system-wide

2hexed opened the pull request:
currently, unlike Firefox, Chromium (and its respective forks) can only be installed via `systemPackages`, since we already handle a lot of chromium and chrome policies/configurations on a system-wide level, it would be nice to have an option to `programs.chromium.enable` and `programs.chromium.package` as well which installs the specified chromium package (browser) system-wide


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

2hexed committed:
nixos/chromium: simple option to install chromium system-wide

currently, unlike Firefox, Chromium (and its respective forks) can only be installed via `systemPackages`, since we already handle a lot of chromium and chrome policies/configurations on a system-wide level, it would be nice to have an option to `programs.chromium.enable` and `programs.chromium.package` as well which installs the specified chromium package (and browser) system-wide

2hexed committed:
Update chromium.nix

syntax error fix

networkException reviewed (changes requested):
Please see https://github.com/NixOS/nixpkgs/pull/334726 for why this isn't the current behavior. This should not be merged as is

2hexed commented:
> Please see #334726 for why this isn't the current behavior. This should not be merged as is

i see, however this PR *only* installs the browser if `.package` is set to a package, leaving the default behavior of the `programs.chromium` module intact

however if its still isn't applicable perhaps we could add an option other than `.enable` to install (and enable) the chromium browser? Perhaps something like `.enableBrowser`, How does that sound to you?

networkException commented:
> i see, however this PR only installs the browser if .package is set to a package, leaving the default behavior of the programs.chromium module intact

I don't think thats the case, the option isn't nullable and the default is set to vanilla chromium

2hexed committed:
nulled default package to keep the default behaviour intact

2hexed committed:
removing types.nullOr

2hexed commented:
hi sorry for the late response but i had the value set to `null` by default so that we can still keep the default-behaviour of the `programs.chromium` as it is :)

SuperSandro2000 commented:
We can just make a changelog entry to announce that breaking change and then we should be good.

SuperSandro2000 commented on nixos/modules/programs/chromium.nix:
```diff
@@ -128,40 +138,44 @@ in
+    environment = {
+      systemPackages = lib.mkIf (cfg.enable && cfg.package != null) [ cfg.package ];
+
+      etc = lib.mkIf cfg.enable {
```
```suggestion
      # Clean those up and only place the relevant policy files
      etc = lib.mkIf cfg.enable {
```

SuperSandro2000 reviewed (commented):
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

2hexed committed:
Merge branch 'NixOS:master' into patch-6

2hexed committed:
nixos/chromium: refactor

In this, I've implemented a few module options as well as variables to
handle etc files and the type of the browser that user specifies; please
let me know if there are any changes required; I'll do my best to
help

2hexed commented:
> I don't think this approach is flexible enough. I suggest providing separate option for each browser whether configure it, along with a "packages" option that allows users to install multiple browsers.

I can't tell if I understand you correctly but `programs.chromium` option is specifically **only** to manage `chrome`, `chromium` (and their derivatives); similar to [`programs.firefox.enable`](https://search.nixos.org/options?channel=24.11&from=0&size=50&sort=relevance&type=packages&query=programs.firefox.enable) including other browser-related (or browser-specific) options.

I hope I'm interpreting "allows users to install multiple browsers" correctly; because it seems unrelated to `programs.chromium` option, not-so-easy to implement if we really think about it (given that we also need to handle config and other browser-specific/policy-related files). 

Besides, if a user really wishes to have multiple web browsers then there are two choices (from what I know),
1. Make use of `programs.<web-browser>.enable` for each browser that they wish to install.
2. We have [`users.users.<name>.packages`](https://search.nixos.org/options?channel=24.11&show=users.users.%3Cname%3E.packages&from=0&size=50&sort=relevance&type=packages&query=users.users.) and [`environment.systemPackages`](https://search.nixos.org/options?channel=24.11&show=environment.systemPackages&from=0&size=50&sort=relevance&type=packages&query=environment.systemPackages)

If not that, users are always allowed to request new modules/packages to be added to nixpkgs and NixOS, eventually someone will end up implementing them :D

midirhee12 commented:
Just responding to a few of the comments here... 

The option in its current form does not do what the option name reads: to enable the program chromium. Every other `programs.*` option installs the actual package corresponding to the name as it reads. This option is the only exception.

So no, I do not think there should be an `enableBrowser` option to further cement the obtuse naming here.

If anything, this option should be repurposed to install chromium. And either the chromium policies are set up as well through the same option or a separate option like `programs.chromium.policies.enable`. To be fair, most `programs.*` will do a little more than just install the package to make it set up properly. So I can see how setting up policies could be done in the existing `programs.chromium.enable` option as long as it installs chromium. And I can also see the reality of having a separate policies option if this really isn't the de-facto case orsane defaulf for most users of chromium.

eclairevoyant reviewed (changes requested):
* This is missing a release notes entry
* This is missing a proper error for the renamed options
* ~~Per https://github.com/NixOS/nixpkgs/issues/390768 this should not be merged before May 15 (branchoff), as it's a breaking change, and non-critical breaking changes were restricted after April 9~~

2hexed commented:
thank you for the feedback, i'll try to work on those issues however i have two questions 

1. how do you add release notes and manual(s) entries?
2. proper error as in error messages? if yes then i can fix that

2hexed committed:
add comment about package set to null

2hexed commented on nixos/modules/programs/chromium.nix:
```diff
@@ -15,14 +15,84 @@ let
+
+      package = lib.mkOption {
+        type = lib.types.package;
+        default = null;
```
done :)

2hexed committed:
wrap whole config inside lib.mkIf followed by cfg.enable

2hexed commented on nixos/modules/programs/chromium.nix:
```diff
@@ -122,47 +192,35 @@ in
 
   ###### implementation
 
   config = {
```
done :)

JohnRTitor reviewed (commented):
Please rebase to remove merge commits.

```
git fetch upstream
git rebase upstream/master
```

And then squash the commits such that:

there's one commit adding the package option,
another for the policies option commit

networkException reviewed (changes requested):
Blocking as maintainer. This doesn't properly address the issues previously brought up in https://github.com/NixOS/nixpkgs/pull/334726.

Specifically, with this change it is no longer possible to use policies without specifying a package and having it added to the system closure.

Moreover, the policies that users currently set using `extraOpts` will be discarded by this change and they will not get warned that `package` needs to be set.

midirhee12 commented:
> Blocking as maintainer. This doesn't properly address the issues previously brought up in #334726.

No, and it was detailed in a comment how this was wrong even. Your change made zero sense and should have never been merged.

networkException commented:
I fear that you are underestimating the complexity of a change like this.

> This way users won't need to rely on home manager or other helpers to configure just a simple browser

Please take a look at the scope of the home-manager module at https://github.com/nix-community/home-manager/blob/master/modules/programs/chromium.nix. It has various options that operate inside the user's local `~/.config/chromium` directory, the NixOS module should never interact with that.

For changing the command like arguments (`chromium.override { commandLineArgs = []; }`) and various other customizations it is furthermore necessary to modify the package used.

Some other customizations are however *only* available through a policy contained in `/etc`, as such not in scope of the home-manager module either.

Enforcing a package to be set in the NixOS module for policies to work **breaks** these setups. It needs to be possible to set policies without putting a package in the system closure. Users shouldn't need to do tricks to not have a chromium package in their system and possible another, customized chromium package in their home closure.

For more issues and edge cases changes to this module uncover please refer to #334726

> I think we can add release notes entry to show the breaking changes or something similar

A release entry is good to convey changes, however users should be informed about breaking changes affecting their configuration at evaluation time. Assertions and warnings are the right tool to change the behavior of options here.

> Your change made zero sense and should have never been merged.

You are referring a pull requesting *reverting* changes to the module that caused breakage. Please don't label changes as not making sense that easily.

midirhee12 commented:
@networkException Yes, it reverted changes that needed not be reverted. You claimed it was breaking when the solution was the thing you removed, the `.package` option like in every other `programs.` option. You should have waited for more review. And now you're overcomplicating this one.

2hexed commented:
> I fear that you are underestimating the complexity of a change like this.
Perhaps but looking at previous PRs and Issues (changes get merged and reverted again and again which is super confusing and unclear - The last change I believe was made by you simply because the module was offering a package option which is what everyone wants I believe but eventually get reverted and it's unclear *why*) relating to enable the Chromium browser (and it's alternatives), we are going to have to implement that at some point right? So I say why not now?

> Please take a look at the scope of the home-manager module at https://github.com/nix-community/home-manager/blob/master/modules/programs/chromium.nix. It has various options that operate inside the user's local ~/.config/chromium directory, the NixOS module should never interact with that.
I hope I'm not being mean or anything but I don't understand the point of Home Manager (and all the other helpers that exist) and I don't use flakes for system configuration at all so I can't tell if I understand those options.

Yes, Home Manager might have several options to configure chromium that operate in the user's home but so does firefox, right? Do the Nix module ([`programs.firefox`](https://search.nixos.org/options?channel=24.11&from=0&size=50&sort=relevance&type=packages&query=programs.firefox)) **only** provide Firefox policies? No, because it seems reasonable to add Browser-specific options. And in future they will be adding every other browser to their flake modules (or whatever/however they offer those modules/options) but this is pointless because we're not working with user's home at all.

> Some other customizations are however only available through a policy contained in /etc, as such not in scope of the home-manager module either.
Indeed, and that is exactly why I'm suggesting these changes to the module

> Enforcing a package to be set in the NixOS module for policies to work breaks these setups. It needs to be possible to set policies without putting a package in the system closure. Users shouldn't need to do tricks to not have a chromium package in their system and possible another, customized chromium package in their home closure.
Right, if we go this way then we should completely remove the [`programs.firefox`](https://search.nixos.org/options?channel=24.11&from=0&size=50&sort=relevance&type=packages&query=programs.firefox), why you might ask, Well, Because the helpers provide it already! :D

> A release entry is good to convey changes, however users should be informed about breaking changes affecting their configuration at evaluation time. Assertions and warnings are the right tool to change the behavior of options here.
I believe assertions and warnings are doable (or something similar) although I'll need some help implementing alongside current changes.

In the end, I believe we should work on how we can make NixOS (and it's internal modules, services, etc) better instead of relying on helpers because there are extra steps that users need to take in order to maintain compatibility along with enabling whatever the helper requires, that is if we're going to support something then either we should support/implement it fully or don't support/implement it at all. :D

New-comers, like me (and/or the users who are just getting started with NixOS) are going to find it complex (possibly unclear as to why even enable/setup them, they might think, is NixOS itself isn't enough?) enabling helpers, and eventually, they will end up complicating their setup which will be troublesome for both the user and us and it might make them think if Nix (and NixOS) is really worth going through all that trouble :(

eclairevoyant commented:
> 1. how do you add release notes and manual(s) entries?

See the PR checklist :)

> 2. proper error as in error messages?

https://github.com/NixOS/nixpkgs/blob/ff74c3e4449f7b15c73c6e0662fe0dbae810c616/nixos/README.md?plain=1#L49-L50

There are hundreds of examples of using this in nixpkgs if you need examples.

midirhee12 commented:
@2hexed Yeah, in addition to what you are saying, if we *really* were to go with the convention of not installing the package then we'd have to revert all of [these](https://search.nixos.org/options?channel=24.11&sort=relevance&type=packages&query=programs.*.enable) as well for consistency.

networkException commented:
This isn't a discussion. We aren't breaking existing functionality and people's config like this.

If you don't *want* to understand home-manager and the issues I brought up then this won't progress. I don't have the time to work on this myself at the moment nor for endless discussions

eclairevoyant commented:
@networkException nothing will break if `mkRenamedOptionModuleWith` is used.

2hexed commented:
> This isn't up for discussion. We aren't breaking existing functionality and people's config like this.
> 
> If you don't _want_ to understand home-manager and the issues I brought up then this won't progress. I don't have the time to work on this myself at the moment nor for endless discussions

I myself have been having a hard time with life lately which is why I've been inactive with this PR, and what you're saying is completely understandable, Take your time off until you're ready to work again, in the meantime we all are here to help! Sending positive vibes to you :DD!

midirhee12 commented:
> This isn't up for discussion. We aren't breaking existing functionality and people's config like this.
> 
> If you don't _want_ to understand home-manager and the issues I brought up then this won't progress. I don't have the time to work on this myself at the moment nor for endless discussions

1. Who made you the sole contributor? This isn't even your PR. 
2. Who made you the sole person to decide?
3. It is up for discussion. In fact, it has been in the process of being discussed already for the last 3 weeks. I'd say it's very much being discussed with quite a few messages.
4. Breaking changes are fine. They just have to occur when they should. For example, your PR was also a breaking change that broke many people's configs.

eclairevoyant commented:
This does not need to be a breaking change and this is needlessly antagonistic. Please focus on the code if you want this change to move forward.

midirhee12 commented:
There's also the option to check for release, like:

```nix
lib.versionAtLeast config.system.nixos.release
```

2hexed commented:
hello @eclairevoyant :)) so I read about the lib function that you mentioned (`mkRenamedOptionModuleWith`) and I understand how to work with it but one little thing that I don't understand is that, for which options do I use it for, because I've only refactored the existing options and I haven't renamed any so I don't know (except for implementing a new `programs.chromium.policies` option; I would assume, use it to rename `.extraOpts` with `.policies` but I still wanted to confirm)

2hexed commented:
Reopening with the hopes that this gets merged with our 25th release or later ^-^

2hexed commented:
Hey guys, sorry if this is annoying but could we please discuss about getting this merged? It's okay if you guys don't want to accept my PR if that's the case. I (and I'm sure a lot of other people as well) would want to have this module/option in NixOS. I understand there are more important things you guys need to maintain but I'm willing to do my best to help with whatever I can. :)

eclairevoyant commented:
Per my prior review, https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md still needs to be updated.

2hexed committed:
nixos/chromium: add chromium module

This commit adds an entry to release notes about chromium being available system-wide.

2hexed committed:
Merge branch 'master' into patch-6

JohnRTitor commented:
@networkException @emilylange Are you satisfied with the current state of the option? The package option is nullable and set to null by default, only after configuring it will install the browser, as far as I can see.

JohnRTitor reviewed (commented):
Please run `git fetch upstream` and `git rebase upstream/master` to rebase.

2hexed committed:
nixos/chromium: entry description update

emilylange commented on nixos/modules/programs/chromium.nix:
```diff
@@ -15,14 +15,85 @@ let
+
+  mergedExtraOpts = browserPolicies // cfg.extraOpts;
+  browserPolicies = cfg.policies.${browserType} or { };
+  selectedBrowserConfig = browserEtcConfig.${browserType} or { };
```
When is also broken when using `null` as `cfg.package`:

```
error: expected a string but found null: null
at ./nixpkgs/nixos/modules/programs/chromium.nix:76:46:
    75|   browserPolicies = cfg.policies.${browserType} or { };
    76|   selectedBrowserConfig = browserEtcConfig.${browserType} or { };
      |                                              ^
    77| in
```

emilylange reviewed (changes requested):
This module, even long before this PR here, has some fundamental issues.

To quote my co-maintainer on this[^1]:

> More design is needed before having policy and package options coexist properly in this module.

This module is in the odd position of doing too much already but also not enough.

It installs policies for multiple `chromium`-based browsers unconditionally, and now it's supposed to install the browsers itself too.

And `brave` should have never been added to this mess. It should have been its own module. We could argue the same for `google-chrome`.

They are all different browsers, they all expect different paths for their policies and their policies itself also differ in features.

This module is a mess, it always has been. But this PR isn't the solution, it only makes the mess worse. The blast radius of this is just very unfortunate.

We should focus on implementing #268834 for `brave` first, and then arguably do the same for `google-chrome`. There will be breaking changes, for sure. But moving `brave` and potentially `google-chrome` out of `programs.chromium` and into their own `programs.brave` and `programs.google-chrome` or whatever, is needed to solve on of the fundamental issues this module has.

Sorry if that's not what you wanted to hear.

Blocking.

[^1]: https://github.com/NixOS/nixpkgs/pull/334726
